### PR TITLE
35% faster completion startup (on ~8300 files)

### PR DESCRIPTION
**Pull request by zot**
Caching: make process-display-format cache its value
SQL: add index to DB, make SQL compute objects for query-nodes.
  Note that there is a dependency between the org-roam-node def and
  org-roam-query-nodes. This is indicated at both places in the code.

### CODE REVIEW

**Comment by zot** [comment]
This adds two indexes to the database (on nodes.id and aliases.node_id) that speeds up the join quite a bit.

I did not update `org-roam-db--init` though -- it just does `create index if not exists` before the completion query. This is safe and takes negligible time but it's probably good to add it to the database initialization.

File: org-roam.el [hunk 1]
```diff
@@ -617,28 +618,39 @@ Throw an error if multiple choices exist."
...
+        (org-roam--sql (org-roam-db) "
+    create index if not exists nodes_id on nodes (id);
+    create index if not exists aliases_id on aliases (node_id);"))))
+
+;; THIS MUST BE IN SYNC WITH ORG-ROAM-NODE DEF!!!
+(defun org-roam-query-nodes ()
+  "Get nodes from database"
+  (org-roam--db-prep)
+  (mapcar #'car (org-roam--sql (org-roam-db) "select '#s(org-roam-node ' || id || ' ' || file || ' nil ' || pos || ' nil nil nil nil ' || title || ' ' || properties || ' ' || ifnull(olp, 'nil') || ' nil nil nil' || ')' from nodes union select '#s(org-roam-node ' || node_id || ' ' || nodes.file || ' nil ' || pos || ' nil nil nil nil ' || alias || ' ' || properties || ' ' || ifnull(olp, 'nil') || ' nil nil nil' || ')' from aliases left join nodes on node_id = id;")))
+
```
Comment (jethrokuan) [inline org-roam.el:643]: These aren't necessary anymore on latest commit.

File: org-roam.el [hunk 2]
```diff
@@ -617,28 +618,39 @@ Throw an error if multiple choices exist."
...
+        (setq org-roam--db-prepped t)
+        (org-roam--sql (org-roam-db) "
+    create index if not exists nodes_id on nodes (id);
+    create index if not exists aliases_id on aliases (node_id);"))))
+
+;; THIS MUST BE IN SYNC WITH ORG-ROAM-NODE DEF!!!
+(defun org-roam-query-nodes ()
+  "Get nodes from database"
+  (org-roam--db-prep)
+  (mapcar #'car (org-roam--sql (org-roam-db) "select '#s(org-roam-node ' || id || ' ' || file || ' nil ' || pos || ' nil nil nil nil ' || title || ' ' || properties || ' ' || ifnull(olp, 'nil') || ' nil nil nil' || ')' from nodes union select '#s(org-roam-node ' || node_id || ' ' || nodes.file || ' nil ' || pos || ' nil nil nil nil ' || alias || ' ' || properties || ' ' || ifnull(olp, 'nil') || ' nil nil nil' || ')' from aliases left join nodes on node_id = id;")))
```
Comment (jethrokuan) [inline org-roam.el:642]: Can we profile how much perf improvements come from manually constructing the query like this?

If the bulk of the perf improvements come from the indexes on the alias tables for the joins, then I'm leaning towards keeping query as is to keep the code clean.

The caching of the display format looks much more acceptable.

Comment (zot) [reply org-roam.el:642]: Sure, I can check what difference it makes

Comment (zot) [reply org-roam.el:642]: It's a good thing you asked that. You are correct that the index is what makes the difference. In fact, besides being harder to read and maintain, the inlined SQL version is actually around 7% slower than the original! I'm guessing that SQLite's string concatenation isn't as well implemented as Emacs'.

**Comment by zot** [comment]
Hmm, I smashed my branch and there's a conflict now.

I'll close this and open another pull request.

**Pull request closed without merging**